core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java

```diff
@@ -0,0 +1,295 @@
+          sb.append((char) c);
+        else
+          log.debug("Dropping non printable char: \\x{}", Integer.toHexString(c));
+      }
+      return sb.toString();
+    }
+  }
+
+  /**
+   * Scan the files for indexed keys since it is more efficient than a full file scan.
+   */
+  private TreeSet<String> getIndexKeys(AccumuloConfiguration accumuloConf, Configuration hadoopConf,
+      FileSystem fs, List<Path> files, boolean base64encode) throws IOException {
+    TreeSet<String> indexKeys = new TreeSet<>();
+    List<SortedKeyValueIterator<Key,Value>> readers = new ArrayList<>(files.size());
+    List<FileSKVIterator> fileReaders = new ArrayList<>(files.size());
+
+    for (Path file : files) {
+      FileSKVIterator reader = FileOperations.getInstance().newIndexReaderBuilder()
+          .forFile(file.toString(), fs, hadoopConf, CryptoServiceFactory.newDefaultInstance())
+          .withTableConfiguration(accumuloConf).build();
+      readers.add(reader);
+      fileReaders.add(reader);
+    }
+    var iterator = new MultiIterator(readers, true);
+    while (iterator.hasTop()) {
+      Key key = iterator.getTopKey();
+      indexKeys.add(encode(base64encode, key.getRow()));
+      iterator.next();
+    }
+
+    for (var r : fileReaders) {
+      r.close();
+    }
+
+    log.debug("Got {} splits from indices of {}", indexKeys.size(), files);
+    return indexKeys;
+  }
+
+  /**
+   * Get number of splits based on requested size of split. A splitSize = 0 returns all keys.
+   */
+  private TreeSet<String> getSplitsBySize(AccumuloConfiguration accumuloConf,
+      Configuration hadoopConf, List<Path> files, FileSystem fs, long splitSize,
+      boolean base64encode) throws IOException {
+    TreeSet<String> splits = new TreeSet<>();
+    List<SortedKeyValueIterator<Key,Value>> readers = new ArrayList<>(files.size());
+    List<FileSKVIterator> fileReaders = new ArrayList<>(files.size());
+    SortedKeyValueIterator<Key,Value> iterator;
+
+    long currentSplitSize = 0;
+    long totalSize = 0;
+    for (Path file : files) {
+      FileSKVIterator reader = FileOperations.getInstance().newScanReaderBuilder()
+          .forFile(file.toString(), fs, hadoopConf, CryptoServiceFactory.newDefaultInstance())
+          .withTableConfiguration(accumuloConf).overRange(new Range(), Set.of(), false).build();
+      readers.add(reader);
+      fileReaders.add(reader);
+    }
+    iterator = new MultiIterator(readers, false);
+    iterator.seek(new Range(), Collections.emptySet(), false);
+    while (iterator.hasTop()) {
+      Key key = iterator.getTopKey();
+      Value val = iterator.getTopValue();
+      int size = key.getSize() + val.getSize();
+      currentSplitSize += size;
+      totalSize += size;
+      if (currentSplitSize > splitSize) {
+        splits.add(encode(base64encode, key.getRow()));
+        currentSplitSize = 0;
+      }
+      iterator.next();
+    }
+    for (var r : fileReaders) {
+      r.close();
+    }
```

Comment:
This could be in a finally block in case there are any exceptions above after the readers are created.

Reply:
Fixed in 7513858